source/logrepl/snapshot.go

```diff
@@ -0,0 +1,221 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logrepl
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/conduitio/conduit-connector-postgres/pgutil"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+
+	sq "github.com/Masterminds/squirrel"
+	"github.com/jackc/pgtype"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+)
+
+var psql = sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
+
+const actionSnapshot string = "snapshot"
+
+type SnapshotConfig struct {
+	URI          string
+	SnapshotName string
+	Table        string
+	Columns      []string
+	KeyColumn    string
+}
+
+type SnapshotIterator struct {
+	config SnapshotConfig
+
+	pool *pgxpool.Pool
+	rows pgx.Rows
+	tx   pgx.Tx
+
+	internalPos int64
+}
+
+func NewSnapshotIterator(ctx context.Context, cfg SnapshotConfig) (*SnapshotIterator, error) {
+	s := &SnapshotIterator{
+		config: cfg,
+	}
+
+	err := s.attachPool(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to attach pool: %w", err)
+	}
+
+	err = s.startTransaction(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to acquire snapshot transaction: %w", err)
+	}
+
+	err = s.loadRows(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load rows: %w", err)
+	}
+
+	return s, nil
+}
+
+func (s *SnapshotIterator) attachPool(ctx context.Context) error {
+	pool, err := pgxpool.Connect(ctx, s.config.URI)
+	if err != nil {
+		return err
+	}
+	s.pool = pool
+	return nil
+}
+
+func (s *SnapshotIterator) loadRows(ctx context.Context) error {
+	snapshotTx := fmt.Sprintf("SET TRANSACTION SNAPSHOT '%s';", s.config.SnapshotName)
+	_, err := s.tx.Query(ctx, snapshotTx)
```

Comment:
We should use `Exec` instead of `Query`, then we don't have to worry about closing the rows (that's what's missing right now and is causing the conn busy error).